documentation/secure_storage_rpmb.md

```diff
@@ -0,0 +1,108 @@
+# RPMB Secure Storage
+
+## Introduction
+
+There are currently two secure storage implementations in OP-TEE:
+
+- The first one relies on the normal world (REE) file system. It is described in
+[secure_storage.md](secure_storage.md) and is the default implementation.
+- The second one makes use of the Replay Protected Memory Block (RPMB) partition
+of an eMMC device, and is enabled by setting `CFG_RPMB_FS=y`.
+
+It is not currently possible to use the normal world filesystem and the RPMB
+implementations simultaneously. This document describes the RPMB implementation.
+The architecture is depicted below.
+
+```
+            NORMAL WORLD           :            SECURE WORLD
+                                   :
+ U        tee-supplicant           :        Trusted application
+ S           (rpmb.c)              :        (secure storage API)
+ E         ^          ^            :                  ^
+ R         |          |            :~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~
+ ~~~~~~~ ioctl ~~~~~~~|~~~~~~~~~~~~:                  v
+ K         |          |            :               OP-TEE
+ E         v          v            :         (tee_svc_storage.c)
+ R    eMMC driver   OP-TEE driver  : (tee_rpmb_fs.c, tee_fs_key_manager.c)
+ N                           ^     :            (tee_rpmb.c)
+ E                           |     :                  ^
+ L                           |     :                  |
+                             |     :                  |
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~
+                             v                        v
+                           Secure monitor / EL3 firmware
+```
+
+## The Secure Storage API
+
+This part is common with the REE-based filesystem. The interface between the
+system calls in [core/tee/tee_svc_storage.c](../core/tee/tee_svc_storage.c) and
+the RPMB filesystem is the **tee_file_operations** structure `tee_file_ops`.
+
+## The RPMB filesystem
+
+The FS implementation is mainly in [core/tee/tee_rpmb_fs.c](../core/tee/tee_rpmb_fs.c).
+
+The RPMB partition is divided in three parts:
+
+- The first 128 bytes are reserved for partition data (struct
+**rpmb_fs_partition**).
+- At offset 512 is the File Allocation Table (FAT). It is an array of
```

Comment:
I double checked the GP spec, and indeed the requirement is quite strict: "TEE_WriteObjectData() [...] Writing in a data stream is atomic; either the entire operation completes successfully or no write is done." so I'll change the if clause. Basically, we can update the data in place only if the update spans at most `rel_wr_blkcnt` blocks.